Add downloader module

qiaozhigang opened the pull request:
Add downloader module

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/seeleteam/go-seele/96)
<!-- Reviewable:end -->

WTer commented:
Review status: 0 of 6 files reviewed at latest revision, all discussions resolved.

---

*[seele/define.go, line 39 at r1](https://beta.reviewable.io/reviews/seeleteam/go-seele/96#-L8fCmtBBpZ86YkTZ6ZN:-L8fCmtBBpZ86YkTZ6ZO:b13agqv) ([raw file](https://github.com/seeleteam/go-seele/blob/55821a3a4026d2692ca553095ecc09b8cc83fd87/seele/define.go#L39)):*
> getBlockHeadersData

get represents action, you will add some functions on this struct, so you'd better to remove 'get'.

---

*[seele/download/downloader.go, line 23 at r1](https://beta.reviewable.io/reviews/seeleteam/go-seele/96#-L8fDULe0FeEq1y_rUUq:-L8fDULe0FeEq1y_rUUr:b-ax0ell) ([raw file](https://github.com/seeleteam/go-seele/blob/55821a3a4026d2692ca553095ecc09b8cc83fd87/seele/download/downloader.go#L23)):*
> 0x81

why started from this number, some hints?

---

*[seele/download/downloader.go, line 46 at r1](https://beta.reviewable.io/reviews/seeleteam/go-seele/96#-L8fDoKXCMeavQ3iPn0V:-L8fDoKXCMeavQ3iPn0W:b-wqyg63) ([raw file](https://github.com/seeleteam/go-seele/blob/55821a3a4026d2692ca553095ecc09b8cc83fd87/seele/download/downloader.go#L46)):*
> match!

remove the last char '!'

---

*[seele/download/downloader.go, line 47 at r1](https://beta.reviewable.io/reviews/seeleteam/go-seele/96#-L8fDipC63AwH_G8zQkb:-L8fDipC63AwH_G8zQkc:bcutqk7) ([raw file](https://github.com/seeleteam/go-seele/blob/55821a3a4026d2692ca553095ecc09b8cc83fd87/seele/download/downloader.go#L47)):*
> recved

typo: received

---

*[seele/download/downloader.go, line 51 at r1](https://beta.reviewable.io/reviews/seeleteam/go-seele/96#-L8fDv6L7Zkz5IFkkqM9:-L8fDv6L7Zkz5IFkkqMA:bbgp4r5) ([raw file](https://github.com/seeleteam/go-seele/blob/55821a3a4026d2692ca553095ecc09b8cc83fd87/seele/download/downloader.go#L51)):*
>  chan
indent other some members, add more blanks before it

---

*[seele/download/downloader.go, line 75 at r1](https://beta.reviewable.io/reviews/seeleteam/go-seele/96#-L8fEAyU84GAf5bX0n_F:-L8fEAyU84GAf5bX0n_G:bihfqer) ([raw file](https://github.com/seeleteam/go-seele/blob/55821a3a4026d2692ca553095ecc09b8cc83fd87/seele/download/downloader.go#L75)):*
> if localTD.Cmp(td) >= 0 {

nice to add comments to explain, such as the local chain is updated or longer chain

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/seeleteam/go-seele/96)*
<!-- Sent from Reviewable.io -->

lucusfly commented:
Review status: 0 of 6 files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[seele/define.go, line 39 at r1](https://beta.reviewable.io/reviews/seeleteam/go-seele/96#-L8fCmtBBpZ86YkTZ6ZN:-L8fofvEBbWA7QXukCM_:brsrvaj) ([raw file](https://github.com/seeleteam/go-seele/blob/55821a3a4026d2692ca553095ecc09b8cc83fd87/seele/define.go#L39)):*
<details><summary><i>Previously, WTer wrote…</i></summary><blockquote>

> getBlockHeadersData

get represents action, you will add some functions on this struct, so you'd better to remove 'get'.
</blockquote></details>

suggest to use blockHeaderQuery

---

*[seele/define.go, line 41 at r1](https://beta.reviewable.io/reviews/seeleteam/go-seele/96#-L8fpDh92FPcoNy0w7ki:-L8fp1Sj8CIgUJZO6Hxa:b-v794cw) ([raw file](https://github.com/seeleteam/go-seele/blob/55821a3a4026d2692ca553095ecc09b8cc83fd87/seele/define.go#L41)):*
> ```Go
> type getBlockHeadersData struct {
> 	Hash    common.Hash // Block hash from which to retrieve headers (excludes Number)
> 	Number  uint64      // Block hash from which to retrieve headers (excludes Hash)
> ```

please add comments when use hash and when use number

---

*[seele/download/downloader.go, line 23 at r1](https://beta.reviewable.io/reviews/seeleteam/go-seele/96#-L8fq-wb6rHVWWF7bP-7:-L8fq-wb6rHVWWF7bP-8:b-22qot0) ([raw file](https://github.com/seeleteam/go-seele/blob/55821a3a4026d2692ca553095ecc09b8cc83fd87/seele/download/downloader.go#L23)):*
> ```Go
> 
> const (
> 	GetBlockHeadersMsg = 0x81
> ```

our reserved control code is 0 to 16. If you need more please change baseProtoCode

---

*[seele/download/downloader.go, line 53 at r1](https://beta.reviewable.io/reviews/seeleteam/go-seele/96#-L8fthPg-kvuJGg39yOp:-L8fthPg-kvuJGg39yOq:b-k3gj5e) ([raw file](https://github.com/seeleteam/go-seele/blob/55821a3a4026d2692ca553095ecc09b8cc83fd87/seele/download/downloader.go#L53)):*
> ```Go
> 	cancelCh chan struct{} // Cancel current synchronising session
> 
> 	masterPeer string               // Identifier of the peer currently being used as the master
> ```

what does masterPeer mean?

---

*[seele/download/downloader.go, line 56 at r1](https://beta.reviewable.io/reviews/seeleteam/go-seele/96#-L8fs29_49o0AS5D96d2:-L8fs29_49o0AS5D96d3:bo8z7gu) ([raw file](https://github.com/seeleteam/go-seele/blob/55821a3a4026d2692ca553095ecc09b8cc83fd87/seele/download/downloader.go#L56)):*
> // 0 for none;
none meaning not start?

---

*[seele/download/downloader.go, line 84 at r1](https://beta.reviewable.io/reviews/seeleteam/go-seele/96#-L8ftHt9EQJupibXv1I6:-L8ftHt9EQJupibXv1I7:behdrqf) ([raw file](https://github.com/seeleteam/go-seele/blob/55821a3a4026d2692ca553095ecc09b8cc83fd87/seele/download/downloader.go#L84)):*
> d.lock.Lock()
> 	if d.syncStatus != statusNone {
> 		d.lock.Unlock()
> 		return errIsSynchronising
> 	}
> 	d.syncStatus = statusPreparing
> 	d.lock.Unlock()

use atomic instead

---

*[seele/download/downloader.go, line 106 at r1](https://beta.reviewable.io/reviews/seeleteam/go-seele/96#-L8fsQW7EMvKgVx8VQzD:-L8fsQW7EMvKgVx8VQzE:b-xguorq) ([raw file](https://github.com/seeleteam/go-seele/blob/55821a3a4026d2692ca553095ecc09b8cc83fd87/seele/download/downloader.go#L106)):*
> ```Go
> }
> 
> func (d *Downloader) synchronise(conn *peerConn, head common.Hash, td *big.Int) (err error) {
> ```

what's the difference between Synchronise and synchronise methods? please find a better name.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/seeleteam/go-seele/96)*
<!-- Sent from Reviewable.io -->

WTer commented:
Review status: 0 of 6 files reviewed at latest revision, 12 unresolved discussions, some commit checks failed.

---

*[seele/define.go, line 25 at r3](https://beta.reviewable.io/reviews/seeleteam/go-seele/96#-L8pYiBk5mBOjCwk4xZ5:-L8pYiBk5mBOjCwk4xZ6:b3m2a3v) ([raw file](https://github.com/seeleteam/go-seele/blob/bbfbd208055e57bbbe915d130a522367086ca581/seele/define.go#L25)):*
> forceSyncInterval

sync what?  nice to add comments.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/seeleteam/go-seele/96)*
<!-- Sent from Reviewable.io -->

lucusfly commented:
Review status: 0 of 6 files reviewed at latest revision, 10 unresolved discussions.

---

*[seele/define.go, line 41 at r1](https://beta.reviewable.io/reviews/seeleteam/go-seele/96#-L8fpDh92FPcoNy0w7ki:-L8q91f51dswmbGmDOUj:btog1pd) ([raw file](https://github.com/seeleteam/go-seele/blob/55821a3a4026d2692ca553095ecc09b8cc83fd87/seele/define.go#L41)):*
<details><summary><i>Previously, qiaozhigang wrote…</i></summary><blockquote>

no need.
</blockquote></details>

why no need? could you explain?

---

*[seele/download/downloader.go, line 53 at r1](https://beta.reviewable.io/reviews/seeleteam/go-seele/96#-L8fthPg-kvuJGg39yOp:-L8q8O5jB4QO17ytb6bQ:bg2gv5p) ([raw file](https://github.com/seeleteam/go-seele/blob/55821a3a4026d2692ca553095ecc09b8cc83fd87/seele/download/downloader.go#L53)):*
<details><summary><i>Previously, qiaozhigang wrote…</i></summary><blockquote>

pls refer to eth if comment is not enough.
</blockquote></details>

better to add more comments, we should not refer any existing project for comments.

---

*[seele/download/downloader.go, line 84 at r1](https://beta.reviewable.io/reviews/seeleteam/go-seele/96#-L8ftHt9EQJupibXv1I6:-L8q8d6T29X2bSsprizw:b-w9s6m4) ([raw file](https://github.com/seeleteam/go-seele/blob/55821a3a4026d2692ca553095ecc09b8cc83fd87/seele/download/downloader.go#L84)):*
<details><summary><i>Previously, qiaozhigang wrote…</i></summary><blockquote>

no need
</blockquote></details>

use atomic instead of lock. these codes only need one line with atomic. and it also make your code more clear.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/seeleteam/go-seele/96)*
<!-- Sent from Reviewable.io -->

qiaozhigang commented:
Review status: 0 of 6 files reviewed at latest revision, 10 unresolved discussions.

---

*[seele/define.go, line 41 at r1](https://beta.reviewable.io/reviews/seeleteam/go-seele/96#-L8fpDh92FPcoNy0w7ki:-L92zx_91HwfQzOdJyLa:b-ea11wo) ([raw file](https://github.com/seeleteam/go-seele/blob/55821a3a4026d2692ca553095ecc09b8cc83fd87/seele/define.go#L41)):*
<details><summary><i>Previously, lucusfly (lucus) wrote…</i></summary><blockquote>

why no need? could you explain?
</blockquote></details>

if have hash when request, then filled with hash
if only have block number, then filled with block number.

---

*[seele/download/downloader.go, line 23 at r1](https://beta.reviewable.io/reviews/seeleteam/go-seele/96#-L8fDULe0FeEq1y_rUUq:-L93-Eyg0z2eEa0jRCn5:b32wtd) ([raw file](https://github.com/seeleteam/go-seele/blob/55821a3a4026d2692ca553095ecc09b8cc83fd87/seele/download/downloader.go#L23)):*
<details><summary><i>Previously, WTer wrote…</i></summary><blockquote>

> 0x81

why started from this number, some hints?
</blockquote></details>

ok

---

*[seele/download/downloader.go, line 84 at r1](https://beta.reviewable.io/reviews/seeleteam/go-seele/96#-L8ftHt9EQJupibXv1I6:-L9307me9nr90TEerEXb:b-n4gvjl) ([raw file](https://github.com/seeleteam/go-seele/blob/55821a3a4026d2692ca553095ecc09b8cc83fd87/seele/download/downloader.go#L84)):*
<details><summary><i>Previously, lucusfly (lucus) wrote…</i></summary><blockquote>

use atomic instead of lock. these codes only need one line with atomic. and it also make your code more clear.
</blockquote></details>

critical object lock must be required, so atomic variable is not need

---

*[seele/download/downloader.go, line 106 at r1](https://beta.reviewable.io/reviews/seeleteam/go-seele/96#-L8fsQW7EMvKgVx8VQzD:-L930PSU27YR98nc5-5B:b32wtd) ([raw file](https://github.com/seeleteam/go-seele/blob/55821a3a4026d2692ca553095ecc09b8cc83fd87/seele/download/downloader.go#L106)):*
<details><summary><i>Previously, lucusfly (lucus) wrote…</i></summary><blockquote>

what's the difference between Synchronise and synchronise methods? please find a better name.
</blockquote></details>

ok

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/seeleteam/go-seele/96)*
<!-- Sent from Reviewable.io -->

lucusfly commented:
Review status: 0 of 6 files reviewed at latest revision, 8 unresolved discussions.

---

*[seele/download/downloader.go, line 84 at r1](https://beta.reviewable.io/reviews/seeleteam/go-seele/96#-L8ftHt9EQJupibXv1I6:-L933sUX0k9AvxIjS-4D:bm4qq1g) ([raw file](https://github.com/seeleteam/go-seele/blob/55821a3a4026d2692ca553095ecc09b8cc83fd87/seele/download/downloader.go#L84)):*
<details><summary><i>Previously, qiaozhigang wrote…</i></summary><blockquote>

critical object lock must be required, so atomic variable is not need
</blockquote></details>

replace it with:
if atomic.CompareAndSwapInt32(&d.syncStatus, statusNone, statusPreparing) {
	d.cancelCh = make(chan struct{})
} else {
        return errIsSynchronising
}

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/seeleteam/go-seele/96)*
<!-- Sent from Reviewable.io -->

lucusfly commented:
Review status: 0 of 6 files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[seele/download/downloader.go, line 23 at r1](https://beta.reviewable.io/reviews/seeleteam/go-seele/96#-L8fq-wb6rHVWWF7bP-7:-L99Qh6q4f8H6w19Xdy0:bwrg3i3) ([raw file](https://github.com/seeleteam/go-seele/blob/55821a3a4026d2692ca553095ecc09b8cc83fd87/seele/download/downloader.go#L23)):*
<details><summary><i>Previously, lucusfly (lucus) wrote…</i></summary><blockquote>

our reserved control code is 0 to 16. If you need more please change baseProtoCode
</blockquote></details>

bigger than 16 is used for protocol msg. your comments are not applied with our implementation.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/seeleteam/go-seele/96)*
<!-- Sent from Reviewable.io -->

WTer commented:
Review status: 0 of 6 files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[seele/download/taskMgr.go, line 55 at r6](https://beta.reviewable.io/reviews/seeleteam/go-seele/96#-L99j60h935H_-wgDXyf:-L99j60h935H_-wgDXyg:b4jne8q) ([raw file](https://github.com/seeleteam/go-seele/blob/7e4e8006adf2ac87ca2ce32b7571049753db902d/seele/download/taskMgr.go#L55)):*
> ```
> master peer
> ```
how can I understand the master peer/node? our whitepaper claims that we don't have master node.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/seeleteam/go-seele/96)*
<!-- Sent from Reviewable.io -->

WTer commented:
Review status: 0 of 6 files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[seele/download/taskMgr.go, line 224 at r6](https://beta.reviewable.io/reviews/seeleteam/go-seele/96#-L99qkRz0T4ro6Wm-6H1:-L99qkRz0T4ro6Wm-6H2:b-guj288) ([raw file](https://github.com/seeleteam/go-seele/blob/7e4e8006adf2ac87ca2ce32b7571049753db902d/seele/download/taskMgr.go#L224)):*
> ```
> t.masterHeaderList = append
> ```
do we need to update t.curNo?

I just see in deliverBlockPreMsg(), you're using t.curNo and t.fromNo

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/seeleteam/go-seele/96)*
<!-- Sent from Reviewable.io -->

qiaozhigang commented:
Review status: 0 of 6 files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[seele/download/downloader.go, line 23 at r1](https://beta.reviewable.io/reviews/seeleteam/go-seele/96#-L8fq-wb6rHVWWF7bP-7:-L9A-4aR4DWhFrj5i6TD:bp1r1j1) ([raw file](https://github.com/seeleteam/go-seele/blob/55821a3a4026d2692ca553095ecc09b8cc83fd87/seele/download/downloader.go#L23)):*
<details><summary><i>Previously, lucusfly (lucus) wrote…</i></summary><blockquote>

bigger than 16 is used for protocol msg. your comments are not applied with our implementation.
</blockquote></details>

Real message code is combined with protoCode and message code,  so logically within a protocol, message code can defined starting with 0.

---

*[seele/download/downloader.go, line 53 at r1](https://beta.reviewable.io/reviews/seeleteam/go-seele/96#-L8fthPg-kvuJGg39yOp:-L99zwjMDoIyyPTFJeT2:b2guq5s) ([raw file](https://github.com/seeleteam/go-seele/blob/55821a3a4026d2692ca553095ecc09b8cc83fd87/seele/download/downloader.go#L53)):*
<details><summary><i>Previously, lucusfly (lucus) wrote…</i></summary><blockquote>

better to add more comments, we should not refer any existing project for comments.
</blockquote></details>

master peer is the best peer.

---

*[seele/download/taskMgr.go, line 55 at r6](https://beta.reviewable.io/reviews/seeleteam/go-seele/96#-L99j60h935H_-wgDXyf:-L99zdwE8l5BmDTsvNNR:bbjaxoe) ([raw file](https://github.com/seeleteam/go-seele/blob/7e4e8006adf2ac87ca2ce32b7571049753db902d/seele/download/taskMgr.go#L55)):*
<details><summary><i>Previously, WTer wrote…</i></summary><blockquote>

> ```
> master peer
> ```
how can I understand the master peer/node? our whitepaper claims that we don't have master node.
</blockquote></details>

master peer is the best peer

---

*[seele/download/taskMgr.go, line 224 at r6](https://beta.reviewable.io/reviews/seeleteam/go-seele/96#-L99qkRz0T4ro6Wm-6H1:-L99uyVv8awJnMmfTQtr:bfmw1q6) ([raw file](https://github.com/seeleteam/go-seele/blob/7e4e8006adf2ac87ca2ce32b7571049753db902d/seele/download/taskMgr.go#L224)):*
<details><summary><i>Previously, WTer wrote…</i></summary><blockquote>

> ```
> t.masterHeaderList = append
> ```
do we need to update t.curNo?

I just see in deliverBlockPreMsg(), you're using t.curNo and t.fromNo
</blockquote></details>

t.curNo is only updated in function run, when block is downloaded already and will be proccessed.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/seeleteam/go-seele/96)*
<!-- Sent from Reviewable.io -->

qiaozhigang commented:
Review status: 0 of 6 files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[seele/download/downloader.go, line 47 at r1](https://beta.reviewable.io/reviews/seeleteam/go-seele/96#-L8fDipC63AwH_G8zQkb:-L9Eacoz7TMHoGwhWK_f:b-896fix) ([raw file](https://github.com/seeleteam/go-seele/blob/55821a3a4026d2692ca553095ecc09b8cc83fd87/seele/download/downloader.go#L47)):*
<details><summary><i>Previously, WTer wrote…</i></summary><blockquote>

> recved

typo: received
</blockquote></details>

Done.

---

*[seele/download/downloader.go, line 51 at r1](https://beta.reviewable.io/reviews/seeleteam/go-seele/96#-L8fDv6L7Zkz5IFkkqM9:-L9Eag-559BeWkzTyGTy:b-896fix) ([raw file](https://github.com/seeleteam/go-seele/blob/55821a3a4026d2692ca553095ecc09b8cc83fd87/seele/download/downloader.go#L51)):*
<details><summary><i>Previously, WTer wrote…</i></summary><blockquote>

>  chan
indent other some members, add more blanks before it
</blockquote></details>

Done.

---

*[seele/download/downloader.go, line 53 at r1](https://beta.reviewable.io/reviews/seeleteam/go-seele/96#-L8fthPg-kvuJGg39yOp:-L9Eagg5BKEcruIews6o:b-896fix) ([raw file](https://github.com/seeleteam/go-seele/blob/55821a3a4026d2692ca553095ecc09b8cc83fd87/seele/download/downloader.go#L53)):*
<details><summary><i>Previously, qiaozhigang wrote…</i></summary><blockquote>

master peer is the best peer.
</blockquote></details>

Done.

---

*[seele/download/downloader.go, line 75 at r1](https://beta.reviewable.io/reviews/seeleteam/go-seele/96#-L8fEAyU84GAf5bX0n_F:-L9EahdS4Xz1Lv_bTVk-:b-896fix) ([raw file](https://github.com/seeleteam/go-seele/blob/55821a3a4026d2692ca553095ecc09b8cc83fd87/seele/download/downloader.go#L75)):*
<details><summary><i>Previously, WTer wrote…</i></summary><blockquote>

> if localTD.Cmp(td) >= 0 {

nice to add comments to explain, such as the local chain is updated or longer chain
</blockquote></details>

Done.

---

*[seele/download/downloader.go, line 84 at r1](https://beta.reviewable.io/reviews/seeleteam/go-seele/96#-L8ftHt9EQJupibXv1I6:-L9EalA-8aOF_8OM1e6C:be38nb5) ([raw file](https://github.com/seeleteam/go-seele/blob/55821a3a4026d2692ca553095ecc09b8cc83fd87/seele/download/downloader.go#L84)):*
<details><summary><i>Previously, lucusfly (lucus) wrote…</i></summary><blockquote>

replace it with:
if atomic.CompareAndSwapInt32(&d.syncStatus, statusNone, statusPreparing) {
	d.cancelCh = make(chan struct{})
} else {
        return errIsSynchronising
}
</blockquote></details>

yes, atomic variable is ok, 
yet considering the simple int variable can also hit the requirment,  I've replaced atomic to int type.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/seeleteam/go-seele/96)*
<!-- Sent from Reviewable.io -->

lucusfly commented:
Review status: 0 of 6 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[seele/download/downloader.go, line 23 at r1](https://beta.reviewable.io/reviews/seeleteam/go-seele/96#-L8fq-wb6rHVWWF7bP-7:-L9EtQLK3nB7SDvrOzyC:bekot0w) ([raw file](https://github.com/seeleteam/go-seele/blob/55821a3a4026d2692ca553095ecc09b8cc83fd87/seele/download/downloader.go#L23)):*
<details><summary><i>Previously, qiaozhigang wrote…</i></summary><blockquote>

Real message code is combined with protoCode and message code,  so logically within a protocol, message code can defined starting with 0.
</blockquote></details>

do you remember that I merged protoCode and message code to only one message code?

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/seeleteam/go-seele/96)*
<!-- Sent from Reviewable.io -->

lucusfly commented:
Better to resolve conflicts with your next iteration. I saw many places that need to resolve.

---

Review status: 0 of 6 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://beta.reviewable.io/reviews/seeleteam/go-seele/96#-:-L9EtkPA4Kvqg_SwEH21:b-n9yf7x)*
<!-- Sent from Reviewable.io -->

qiaozhigang commented:
Review status: 0 of 7 files reviewed at latest revision, 4 unresolved discussions.

---

*[seele/download/downloader.go, line 23 at r1](https://beta.reviewable.io/reviews/seeleteam/go-seele/96#-L8fq-wb6rHVWWF7bP-7:-L9F5Kn2-Qr8Zq6-e0f3:btqeasp) ([raw file](https://github.com/seeleteam/go-seele/blob/55821a3a4026d2692ca553095ecc09b8cc83fd87/seele/download/downloader.go#L23)):*
<details><summary><i>Previously, lucusfly (lucus) wrote…</i></summary><blockquote>

do you remember that I merged protoCode and message code to only one message code?
</blockquote></details>

sure, but you said that would make implementation to be complicated, so leave protoCode messageCode alone.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/seeleteam/go-seele/96)*
<!-- Sent from Reviewable.io -->

qiaozhigang commented:
Review status: 0 of 7 files reviewed at latest revision, 4 unresolved discussions.

---

*[seele/download/downloader.go, line 23 at r1](https://beta.reviewable.io/reviews/seeleteam/go-seele/96#-L8fq-wb6rHVWWF7bP-7:-L9FEsds7Ndnkt-ySWEj:b-896fix) ([raw file](https://github.com/seeleteam/go-seele/blob/55821a3a4026d2692ca553095ecc09b8cc83fd87/seele/download/downloader.go#L23)):*
<details><summary><i>Previously, qiaozhigang wrote…</i></summary><blockquote>

sure, but you said that would make implementation to be complicated, so leave protoCode messageCode alone.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/seeleteam/go-seele/96)*
<!-- Sent from Reviewable.io -->

lucusfly commented:
Review status: 0 of 7 files reviewed at latest revision, 4 unresolved discussions.

---

*[seele/peer.go, line 137 at r9](https://beta.reviewable.io/reviews/seeleteam/go-seele/96#-L9VIdd07n8v5SjSFbBJ:-L9VIdd07n8v5SjSFbBK:b-25hqg4) ([raw file](https://github.com/seeleteam/go-seele/blob/3ee199ed3b9cb3c261e182879fca7c51742f6016/seele/peer.go#L137)):*
> ```Go
> // RequestBlocksByHashOrNumber fetches a batch of blocks corresponding to the
> // specified header query, based on the hash of an origin block.
> func (p *peer) RequestBlocksByHashOrNumber(origin common.Hash, num uint64, amount int) error {
> ```

I already added a similar method. please merge them to one.

---

*[seele/peer.go, line 142 at r9](https://beta.reviewable.io/reviews/seeleteam/go-seele/96#-L9VIJ8jCGwOHKqZw4aX:-L9VIJ8jCGwOHKqZw4aY:b-9sqfjv) ([raw file](https://github.com/seeleteam/go-seele/blob/3ee199ed3b9cb3c261e182879fca7c51742f6016/seele/peer.go#L142)):*
> ```Go
> }
> 
> func (p *peer) sendNewBlockHash(block *types.Block) {
> ```

I already added this method. please use it.

---

*[seele/peer.go, line 146 at r9](https://beta.reviewable.io/reviews/seeleteam/go-seele/96#-L9VINlOCHnQTgy1dO7Q:-L9VINlOCHnQTgy1dO7R:b-9sqfjv) ([raw file](https://github.com/seeleteam/go-seele/blob/3ee199ed3b9cb3c261e182879fca7c51742f6016/seele/peer.go#L146)):*
> ```Go
> }
> 
> func (p *peer) sendNewBlock(block *types.Block, td *big.Int) {
> ```

I already added this method. please use it.

---

*[seele/peer.go, line 150 at r9](https://beta.reviewable.io/reviews/seeleteam/go-seele/96#-L9VIRg89ZOL1lZuaEhv:-L9VIRg89ZOL1lZuaEhw:b-g0n8pf) ([raw file](https://github.com/seeleteam/go-seele/blob/3ee199ed3b9cb3c261e182879fca7c51742f6016/seele/peer.go#L150)):*
> ```Go
> }
> 
> func (p *peer) sendTransactions([]*types.Transaction) error {
> ```

I already added send one transaction method. please change it to send transaction array and use it.

---

*[seele/seeleprotocol.go, line 92 at r9](https://beta.reviewable.io/reviews/seeleteam/go-seele/96#-L9VJCcn8o6tkb5xD2cO:-L9VJCcn8o6tkb5xD2cP:bpbqdqo) ([raw file](https://github.com/seeleteam/go-seele/blob/3ee199ed3b9cb3c261e182879fca7c51742f6016/seele/seeleprotocol.go#L92)):*
> ```
> sp.wg.Wait()
> ```
this will make the stop method stuck. we need find a way to make all the wg.Done() called.

---

*[seele/seeleprotocol.go, line 119 at r9](https://beta.reviewable.io/reviews/seeleteam/go-seele/96#-L9VLuOO-nSxLY7O3Nu1:-L9VLuOO-nSxLY7O3Nu2:b-jhxr64) ([raw file](https://github.com/seeleteam/go-seele/blob/3ee199ed3b9cb3c261e182879fca7c51742f6016/seele/seeleprotocol.go#L119)):*
> ```Go
> 
> // syncTransactions sends pending transactions to remote peer.
> func (sp *SeeleProtocol) syncTransactions(p *peer) {
> ```

I think I already done the sync transaction part. Could you check it?

---

*[seele/seeleprotocol.go, line 123 at r9](https://beta.reviewable.io/reviews/seeleteam/go-seele/96#-L9VL-w86C91qI3xLMXl:-L9VL-w86C91qI3xLMXm:bhzbpv8) ([raw file](https://github.com/seeleteam/go-seele/blob/3ee199ed3b9cb3c261e182879fca7c51742f6016/seele/seeleprotocol.go#L123)):*
> ```Go
> 
> 	//pending, _ := sp.txPool.Pending()
> 	var pending []*types.Transaction //TODO get pending transactions from txPool
> ```

why not use "pending, _ := sp.txPool.Pending()"? Without it, this method is doing nothing.

---

*[seele/seeleprotocol.go, line 147 at r9](https://beta.reviewable.io/reviews/seeleteam/go-seele/96#-L9VMQd21Eg0gi50z5jJ:-L9VMQd21Eg0gi50z5jK:bpsq6ry) ([raw file](https://github.com/seeleteam/go-seele/blob/3ee199ed3b9cb3c261e182879fca7c51742f6016/seele/seeleprotocol.go#L147)):*
> ```Go
> 	}
> 
> 	resultCh <- nil
> ```

this method will stuck here.

---

*[seele/seeleprotocol.go, line 155 at r9](https://beta.reviewable.io/reviews/seeleteam/go-seele/96#-L9VM5nHEdB0KmHLgmf5:-L9VM5nHEdB0KmHLgmf6:b-z0l7bt) ([raw file](https://github.com/seeleteam/go-seele/blob/3ee199ed3b9cb3c261e182879fca7c51742f6016/seele/seeleprotocol.go#L155)):*
> ```Go
> 				break loopOut
> 			}
> 			send(curPos)
> ```

curPos seems not change. this will not send all the transactions succeed.

---

*[seele/download/downloader.go, line 54 at r9](https://beta.reviewable.io/reviews/seeleteam/go-seele/96#-L9VMgui383RhxaVH-qO:-L9VMgui383RhxaVH-qP:bh55z4u) ([raw file](https://github.com/seeleteam/go-seele/blob/3ee199ed3b9cb3c261e182879fca7c51742f6016/seele/download/downloader.go#L54)):*
> ```Go
> type Downloader struct {
> 	cancelCh   chan struct{}        // Cancel current synchronising session
> 	masterPeer string               // Identifier of the best peer
> ```

rename it to bestPeer?

---

*[seele/download/downloader.go, line 135 at r9](https://beta.reviewable.io/reviews/seeleteam/go-seele/96#-L9VPOLF1jfOO9P8ONId:-L9VPOLF1jfOO9P8ONIe:b-v8cmp1) ([raw file](https://github.com/seeleteam/go-seele/blob/3ee199ed3b9cb3c261e182879fca7c51742f6016/seele/download/downloader.go#L135)):*
> ```
> d.tm = tm
> ```

add lock?

---

*[seele/download/peer.go, line 30 at r9](https://beta.reviewable.io/reviews/seeleteam/go-seele/96#-L9VQvwG3rbAwaQ1toGv:-L9VQvwG3rbAwaQ1toGw:b-o0nb75) ([raw file](https://github.com/seeleteam/go-seele/blob/3ee199ed3b9cb3c261e182879fca7c51742f6016/seele/download/peer.go#L30)):*
> ```Go
> type peerConn struct {
> 	peerID         string
> 	peer           Peer
> ```

I didn't find where you use this peer to sendMsg and getMsg. please make sure to use peer.rw to send and get message.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/seeleteam/go-seele/96)*
<!-- Sent from Reviewable.io -->

qiaozhigang commented:
Review status: 0 of 7 files reviewed at latest revision, 15 unresolved discussions.

---

*[seele/peer.go, line 137 at r9](https://beta.reviewable.io/reviews/seeleteam/go-seele/96#-L9VIdd07n8v5SjSFbBJ:-L9YBiqS-cIE114AxOAL:b3kre) ([raw file](https://github.com/seeleteam/go-seele/blob/3ee199ed3b9cb3c261e182879fca7c51742f6016/seele/peer.go#L137)):*
<details><summary><i>Previously, lucusfly (lucus) wrote…</i></summary><blockquote>

I already added a similar method. please merge them to one.
</blockquote></details>

?

---

*[seele/peer.go, line 142 at r9](https://beta.reviewable.io/reviews/seeleteam/go-seele/96#-L9VIJ8jCGwOHKqZw4aX:-L9YBm4I77So9CPPc4Yc:b-rkelob) ([raw file](https://github.com/seeleteam/go-seele/blob/3ee199ed3b9cb3c261e182879fca7c51742f6016/seele/peer.go#L142)):*
<details><summary><i>Previously, lucusfly (lucus) wrote…</i></summary><blockquote>

I already added this method. please use it.
</blockquote></details>

ok, remove this

---

*[seele/peer.go, line 146 at r9](https://beta.reviewable.io/reviews/seeleteam/go-seele/96#-L9VINlOCHnQTgy1dO7Q:-L9YBpGa9fDOE94UkVt5:b-rkelob) ([raw file](https://github.com/seeleteam/go-seele/blob/3ee199ed3b9cb3c261e182879fca7c51742f6016/seele/peer.go#L146)):*
<details><summary><i>Previously, lucusfly (lucus) wrote…</i></summary><blockquote>

I already added this method. please use it.
</blockquote></details>

ok, remove this

---

*[seele/peer.go, line 150 at r9](https://beta.reviewable.io/reviews/seeleteam/go-seele/96#-L9VIRg89ZOL1lZuaEhv:-L9YBrZu1G46WkJgJYRC:bicz9vf) ([raw file](https://github.com/seeleteam/go-seele/blob/3ee199ed3b9cb3c261e182879fca7c51742f6016/seele/peer.go#L150)):*
<details><summary><i>Previously, lucusfly (lucus) wrote…</i></summary><blockquote>

I already added send one transaction method. please change it to send transaction array and use it.
</blockquote></details>

ok remove this

---

*[seele/seeleprotocol.go, line 92 at r9](https://beta.reviewable.io/reviews/seeleteam/go-seele/96#-L9VJCcn8o6tkb5xD2cO:-L9YBv0R8ZtHbvPOrqXc:b32wtd) ([raw file](https://github.com/seeleteam/go-seele/blob/3ee199ed3b9cb3c261e182879fca7c51742f6016/seele/seeleprotocol.go#L92)):*
<details><summary><i>Previously, lucusfly (lucus) wrote…</i></summary><blockquote>

> ```
> sp.wg.Wait()
> ```
this will make the stop method stuck. we need find a way to make all the wg.Done() called.
</blockquote></details>

ok

---

*[seele/seeleprotocol.go, line 123 at r9](https://beta.reviewable.io/reviews/seeleteam/go-seele/96#-L9VL-w86C91qI3xLMXl:-L9Xx5Jl301XK3DhysrQ:b304u3l) ([raw file](https://github.com/seeleteam/go-seele/blob/3ee199ed3b9cb3c261e182879fca7c51742f6016/seele/seeleprotocol.go#L123)):*
<details><summary><i>Previously, lucusfly (lucus) wrote…</i></summary><blockquote>

why not use "pending, _ := sp.txPool.Pending()"? Without it, this method is doing nothing.
</blockquote></details>

pending function has removed from txpool, 
this will be handled in future

---

*[seele/seeleprotocol.go, line 147 at r9](https://beta.reviewable.io/reviews/seeleteam/go-seele/96#-L9VMQd21Eg0gi50z5jJ:-L9YDR1y1v5IpmA-T4_4:b32wtd) ([raw file](https://github.com/seeleteam/go-seele/blob/3ee199ed3b9cb3c261e182879fca7c51742f6016/seele/seeleprotocol.go#L147)):*
<details><summary><i>Previously, lucusfly (lucus) wrote…</i></summary><blockquote>

this method will stuck here.
</blockquote></details>

ok

---

*[seele/seeleprotocol.go, line 155 at r9](https://beta.reviewable.io/reviews/seeleteam/go-seele/96#-L9VM5nHEdB0KmHLgmf5:-L9XxKlY8LEliLu4qm5J:b-2qzknf) ([raw file](https://github.com/seeleteam/go-seele/blob/3ee199ed3b9cb3c261e182879fca7c51742f6016/seele/seeleprotocol.go#L155)):*
<details><summary><i>Previously, lucusfly (lucus) wrote…</i></summary><blockquote>

curPos seems not change. this will not send all the transactions succeed.
</blockquote></details>

curPos = curPos + needSend

---

*[seele/download/downloader.go, line 84 at r1](https://beta.reviewable.io/reviews/seeleteam/go-seele/96#-L8ftHt9EQJupibXv1I6:-L9XxNhq0SNB-0uJMhaw:b-896fix) ([raw file](https://github.com/seeleteam/go-seele/blob/55821a3a4026d2692ca553095ecc09b8cc83fd87/seele/download/downloader.go#L84)):*
<details><summary><i>Previously, qiaozhigang wrote…</i></summary><blockquote>

yes, atomic variable is ok, 
yet considering the simple int variable can also hit the requirment,  I've replaced atomic to int type.
</blockquote></details>

Done.

---

*[seele/download/downloader.go, line 54 at r9](https://beta.reviewable.io/reviews/seeleteam/go-seele/96#-L9VMgui383RhxaVH-qO:-L9Xxhk_DEU4mgsJx12B:b-slf6sq) ([raw file](https://github.com/seeleteam/go-seele/blob/3ee199ed3b9cb3c261e182879fca7c51742f6016/seele/download/downloader.go#L54)):*
<details><summary><i>Previously, lucusfly (lucus) wrote…</i></summary><blockquote>

rename it to bestPeer?
</blockquote></details>

masterpeer is also ok

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/seeleteam/go-seele/96)*
<!-- Sent from Reviewable.io -->

lucusfly commented:
Reviewed 1 of 4 files at r4, 2 of 5 files at r8, 2 of 3 files at r9, 2 of 2 files at r10.
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---



*Comments from [Reviewable](https://beta.reviewable.io/reviews/seeleteam/go-seele/96)*
<!-- Sent from Reviewable.io -->

qiaozhigang commented:
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[seele/download/peer.go, line 30 at r9](https://beta.reviewable.io/reviews/seeleteam/go-seele/96#-L9VQvwG3rbAwaQ1toGv:-L9ZOWXREEjkk5nClyp4:bg2gcz4) ([raw file](https://github.com/seeleteam/go-seele/blob/3ee199ed3b9cb3c261e182879fca7c51742f6016/seele/download/peer.go#L30)):*
<details><summary><i>Previously, lucusfly (lucus) wrote…</i></summary><blockquote>

I didn't find where you use this peer to sendMsg and getMsg. please make sure to use peer.rw to send and get message.
</blockquote></details>

Peer is a interface, seele.peer implements real method.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/seeleteam/go-seele/96)*
<!-- Sent from Reviewable.io -->

qiaozhigang commented:
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[seele/seeleprotocol.go, line 119 at r9](https://beta.reviewable.io/reviews/seeleteam/go-seele/96#-L9VLuOO-nSxLY7O3Nu1:-L9ZfMnr39OLuvBe9WEg:b32wtd) ([raw file](https://github.com/seeleteam/go-seele/blob/3ee199ed3b9cb3c261e182879fca7c51742f6016/seele/seeleprotocol.go#L119)):*
<details><summary><i>Previously, lucusfly (lucus) wrote…</i></summary><blockquote>

I think I already done the sync transaction part. Could you check it?
</blockquote></details>

ok

---

*[seele/download/downloader.go, line 135 at r9](https://beta.reviewable.io/reviews/seeleteam/go-seele/96#-L9VPOLF1jfOO9P8ONId:-L9ZhBBrEh_oKSN1ZAq5:b32wtd) ([raw file](https://github.com/seeleteam/go-seele/blob/3ee199ed3b9cb3c261e182879fca7c51742f6016/seele/download/downloader.go#L135)):*
<details><summary><i>Previously, lucusfly (lucus) wrote…</i></summary><blockquote>

> ```
> d.tm = tm
> ```

add lock?
</blockquote></details>

ok

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/seeleteam/go-seele/96)*
<!-- Sent from Reviewable.io -->

WTer commented:
Reviewed 1 of 4 files at r4, 2 of 5 files at r8, 2 of 3 files at r9, 2 of 2 files at r10.
Review status: :shipit: all files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



*Comments from [Reviewable](https://beta.reviewable.io/reviews/seeleteam/go-seele/96)*
<!-- Sent from Reviewable.io -->
